File: sections/14_core_patterns_mixed_contest/problems/x_subarray_minimum_total/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    // (value, number of suffixes for which value is the minimum)
    vector<pair<long long, long long>> stack;
    long long ending_sum = 0;
    long long answer = 0;

    for (int i = 0; i < n; ++i) {
        long long value;
        cin >> value;
        long long ways = 1;

        while (!stack.empty() && stack.back().first >= value) {
            auto [old_value, old_ways] = stack.back();
            stack.pop_back();
            ending_sum -= old_value * old_ways;
            ways += old_ways;
        }

        stack.push_back({value, ways});
        ending_sum += value * ways;
        answer += ending_sum;
    }

    cout << answer << '\n';
}

File: sections/14_core_patterns_mixed_contest/problems/x_subarray_minimum_total/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    n = data[0]
    values = data[1:]

    # Each pair is (minimum, number of suffixes having that minimum).
    stack: list[tuple[int, int]] = []
    ending_sum = 0
    answer = 0

    for value in values[:n]:
        ways = 1
        while stack and stack[-1][0] >= value:
            old_value, old_ways = stack.pop()
            ending_sum -= old_value * old_ways
            ways += old_ways

        stack.append((value, ways))
        ending_sum += value * ways
        answer += ending_sum

    print(answer)


if __name__ == "__main__":
    main()
